### DISCUSSION TITLE
How do I manually send a password-reset e-mail using the back-end API

### DISCUSSION THREAD

**Discussion by YketD** [swellstores/community#338] [Help] [answered]
## I need help with

Any clarity on how to manually send notifications using the `/notifications` endpoint.

Or in general; how to use the swell API to figure out what's missing / going wrong with requests (for instance through the developer console).

## Context

When a new customer is created, there is a automatic welcome email that's being sent.
Besides that mail, we also want to manually trigger a "forgot-password" email, to push the user to set their password.

I have tried to manually trigger the e-mail (see below for payload), I receive no indication that the request has gone wrong (200 and no body), however there's no e-mail being sent.

Also, I can find no documentation regarding the `notifications` endpoint, so it's very difficult to find a lead on what i'm doing wrong.

### Attempt 1
```
POST :  notifications/send
Payload:
{
   "type":"password-reset",
   "data":{
      "account_id":"{{account_id}}",
      "password_reset_url":"{{url}}"
   }
}
```
Response:
```
array:6 [ // vendor/swellstores/swell-php/lib/Client.php:312
  "$links" => array:2 [
    "record" => array:1 [
      "url" => "/{config.model}/{record_id}"
    ]
    "config" => array:1 [
      "url" => "/:notifications/{template}"
    ]
  ]
  "$url" => "/notifications/send"
  "$collection" => "com.notifications"
  "$data" => null
  "$time" => 11
  "$status" => 200
]
```
## Attempt 2
```
POST :  notifications/send
Payload:
{
   "template":"password-reset",
   "data":{
      "account_id":"{{account_id}}",
      "password_reset_url":"{{url}}"
   }
}
```
Response
```
array:6 [ // vendor/swellstores/swell-php/lib/Client.php:312
  "$links" => array:2 [
    "record" => array:1 [
      "url" => "/{config.model}/{record_id}"
    ]
    "config" => array:1 [
      "url" => "/:notifications/{template}"
    ]
  ]
  "$url" => "/notifications/send"
  "$collection" => "com.notifications"
  "$data" => null
  "$time" => 3
  "$status" => 200
]
```

**Comment by ericingram** [comment, marked as answer]
Appreciate the call-out. We're starting to add documentation for notifications and will post here when it's done.

The right way to force-send notifications for any model looks like this:

```
PUT /model/id
{
  "$notify": "notification_name"
}
```

You can see the available notifications and their respective models and notification names with this request:

```
GET /:notifications
```

For example, to send a password reset email:

```
PUT /accounts/<email>
{
  $notify: 'password-reset',
  password_reset_url, // the URL of the password reset landing page, with {password_reset_key} as a placeholder
  // for example: password_reset_url=https://mystore.com/password-reset?key={password_reset_key}
  password_reset_key, // ideally a random 32+ character string
  password_reset_expired, // a date in the future to expire the reset key
}
```

Specifically for password reset, you might consider using Swell.js which handles most of these parameters by default: https://developers.swell.is/frontend-api/account-recovery#send-a-password-reset-email

**Reply by YketD** [reply]
Great, indeed the email gets send now, thanks a lot for the clear answer. 

I'm now running into a small follow up problem with the answer, which may be better for a separate issue. if so, please notify me and i'll move it to another issue.

 When i send the request, the button in the reset password e-mail links directly to the url provided in the password_reset_url property, however it does not replace the {password_reset_key} part with the reset key.

Could be a bug? it seems to url encode that part. URL in the email source is as follows: `https://example.com/password-reset?key=%7Bpassword_reset_key%7D `

**Reply by papertokyo** [reply]
@YketD Could you share the code you're using for the PUT request to the accounts endpoint?

**Reply by ericingram** [reply]
@YketD sorry I was wrong about that. It also is part of the frontend API `swell.account.recover` method, but doesn't do that in the backend.

Since you're making this request from a secure server I presume, you can just put the key directly in the `url` instead of using the placeholder.

**Reply by YketD** [reply]
Thanks a lot, putting the key directly in the url did the trick 👍
